### ISSUE TITLE
Reaper running too long

### ISSUE THREAD

**Issue by sanjay92gupta** [thelastpickle/cassandra-reaper#1515]
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=74376566)

I see 266/268 segments repaired but last 2 are going on for last 2 days.
On Reaper GUI, I see fail count increased 
Reaper log shows shows this
i.c.s.RepairRunner - Repair amount done 266.0 
INFO   [2024-08-11 16:36:38,457] [cassandra:b8b36e90-5689-11ef-a154-d10bc5a36b53] i.c.s.RepairRunner - Attempting to run new segment... 
INFO   [2024-08-11 16:36:38,474] [cassandra:b8b36e90-5689-11ef-a154-d10bc5a36b53] i.c.s.RepairRunner - All nodes are busy or have too many pending compactions for the remaining candidate segments. 

nodetool compactionstats and nodetool netstats shows some activity but I see that is going on over 2 days so something is not right. Any idea how to debug ?

**Comment by adejanovski** [comment]
You may have some problematic partitions in these two segments.
Check the Reaper logs to see if they reach the timeout or fail on the Cassandra side.
If they reach the timeout, you'll need to raise it maybe (although the adaptive feature should have raised it already).
If they fail on the Cassandra side, you may have some problematic partitions or sstables and need to check the Cassandra logs for more info.

**Comment by sanjay92gupta** [comment]
I don't see ant timeouts. Infact I did increase timeout to 600 on reaper (yml).
I also ran repair for question table using reaper and that ran fine but it took 10 hrs or so but again when running reaper for whole keyspace it just doesn't finish. 
Reaper log shows following
ERROR  [2024-08-13 17:55:31,791] [gn-ncl-nonprod-cassandra:1cb40690-58e5-11ef-96ba-759fa0b871ff:1cb58d35-58e5-11ef-96ba-759fa0b871ff] i.c.s.RepairRunner - Executing SegmentRunner failed 
java.lang.AssertionError: non-leader trying to update repair segment 1cb58d35-58e5-11ef-96ba-759fa0b871ff of run 1cb40690-58e5-11ef-96ba-759fa0b871ff
	at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:1120)
	at io.cassandrareaper.service.SegmentRunner.postpone(SegmentRunner.java:198)
	at io.cassandrareaper.service.SegmentRunner.abort(SegmentRunner.java:213)
	at io.cassandrareaper.service.SegmentRunner.abort(SegmentRunner.java:226)
	at io.cassandrareaper.service.SegmentRunner.processTriggeredSegment(SegmentRunner.java:415)
	at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:313)
	at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:161)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66)
	at java.lang.Thread.run(Thread.java:748)


INFO   [2024-08-13 17:55:41,631] [gn-ncl-nonprod-cassandra:1cb40690-58e5-11ef-96ba-759fa0b871ff:1cb53f1a-58e5-11ef-96ba-759fa0b871ff] i.c.s.RepairRunner - Triggered repair of segment 1cb53f1a-58e5-11ef-96ba-759fa0b871ff via host 10.204.72.214 
INFO   [2024-08-13 17:55:41,631] [gn-ncl-nonprod-cassandra:1cb40690-58e5-11ef-96ba-759fa0b871ff:1cb53f1a-58e5-11ef-96ba-759fa0b871ff] i.c.s.SegmentRunner - Repair for segment 1cb53f1a-58e5-11ef-96ba-759fa0b871ff started, status wait will timeout in 16200000 millis 
INFO   [2024-08-13 17:55:51,620] [gn-ncl-nonprod-cassandra:1cb40690-58e5-11ef-96ba-759fa0b871ff] i.c.s.RepairRunner - Attempting to run new segment... 
INFO   [2024-08-13 17:55:51,634] [gn-ncl-nonprod-cassandra:1cb40690-58e5-11ef-96ba-759fa0b871ff] i.c.s.RepairRunner - All nodes are busy or have too many pending compactions for the remaining candidate segments.

**Comment by sanjay92gupta** [comment]
I have also tried rolling restart cluster but reaper always get stuck 265/268.
Any idea on how to fix this issue ?
Based on reaper.log, I see error
```
ERROR  [2024-08-16 17:28:20,260] [gn-ncl-nonprod-cassandra:67cfc1a0-5be2-11ef-81d8-59649d312db3:67d20baa-5be2-11ef-81d8-59649d312db3] i.c.s.RepairRunner - Executing SegmentRunner failed
java.lang.AssertionError: non-leader trying to update repair segment 67d20baa-5be2-11ef-81d8-59649d312db3 of run 67cfc1a0-5be2-11ef-81d8-59649d312db3
        at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:1120)
        at io.cassandrareaper.service.SegmentRunner.postpone(SegmentRunner.java:198)
        at io.cassandrareaper.service.SegmentRunner.abort(SegmentRunner.java:213)
        at io.cassandrareaper.service.SegmentRunner.abort(SegmentRunner.java:226)
        at io.cassandrareaper.service.SegmentRunner.processTriggeredSegment(SegmentRunner.java:415)
        at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:313)
        at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:161)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66)
        at java.lang.Thread.run(Thread.java:748)

```

**Comment by sanjay92gupta** [comment]
I was able to resolve issue. Issue were some data files for vod_content_reprocess-4752923014a011ed9ebfdfadb9ed3327 table which had very old time stamp. Once I moved/deleted these data files to other backup location and ran repair, it worked fine. So in nutshell some old left over datafiles were causing issue. notetool compact on the problem node also said sst table is corrupt. It looks like nodetool repair didn't complain about corrupt sst table but nodetool compact did mention about sst datafile corruption. I ran repair using reaper also and it went fine.
